server/api/cvrs.py

```diff
@@ -117,15 +128,45 @@ def process():
 
             for row in cvrs:
                 [
-                    _cvr_number,
+                    cvr_number,
                     tabulator_number,
                     batch_id,
                     record_id,
                     imprinted_id,
                     *_,  # CountingGroup (maybe), PrecintPortion, BallotType
                 ] = row[:first_contest_column]
                 interpretations = row[first_contest_column:]
-                db_batch_id = batch_key_to_id[(tabulator_number, batch_id)]
+
+                db_batch_id = batch_key_to_id.get((tabulator_number, batch_id))
+                if not db_batch_id:
+                    close_matches = difflib.get_close_matches(
+                        str((tabulator_number, batch_id)),
+                        (str(batch_key) for batch_key in batch_key_to_id),
+                        n=1,
+                    )
+                    closest_match = (
+                        ast.literal_eval(close_matches[0]) if close_matches else None
+                    )
+
+                    raise UserError(
+                        "Invalid TabulatorNum/BatchId for row with"
+                        f" CvrNumber {cvr_number}: {tabulator_number}, {batch_id}."
+                        " The TabulatorNum and BatchId fields in the CVR file"
+                        " must match the Tabulator and Batch Name fields in the"
+                        " ballot manifest."
+                        + (
+                            (
+                                " The closest match we found in the ballot manifest was:"
+                                f" {closest_match[0]}, {closest_match[1]}."
+                            )
+                            if closest_match
+                            else ""
+                        )
+                        + " Please check your CVR file and ballot manifest thoroughly"
+                        " to make sure these values match - there may be a similar"
+                        " inconsistency in other rows in the CVR file."
```

Comment:
Do we have a check for duplicate rows? Would be nice to call that out explicitly too.

Reply:
Also total rows

Reply:
I don't think it's likely to happen since it's a file generated by their voting system. I'm taking the approach of only building error messages for the errors we've actually seen, since there are tons of ways this file _could_ be wrong. I'm hoping that we can trust the generated files to have the correct format, and discover in practice which ways our assumptions about the format are incorrect.